apps/api/src/view-models/buildViewModelForResource/viewModels/utilities/graph/buildTreeFromNodes/findRoot.spec.ts

```diff
@@ -0,0 +1,63 @@
+import { Category } from '../../../../../../domain/models/categories/entities/category.entity';
+import { DTO } from '../../../../../../types/DTO';
+import findRoot from './findRoot';
+
+describe('findRoot', () => {
+    describe('when given a valid tree', () => {
```

Comment:
It seems like we may want a utility to validate that an array representation of a tree is indeed a tree structure. This may save us in the future.

**Related Question**
How do we ensure that we are maintaining a category tree and not a more general category graph? Where does the responsibility lie?

Reply:
`Category.validateExternalState` validates that every `childID` in `childrenIDs` is an existing Category. However, this does not guarantee that there are no loops.

This raises an important point
- There are some external state rules that cannot be validated within the context of a single model

The rule would be:
- No category is its own descendant

However, we could validate that a single Category model is not responsible for breaking this rule
- categoryRepository.fetchTree(myCategoryId)
- ensure that this sub-tree does not reference `myCategoryId` in any `childrenIDs` property for any nodes.
Let's add [a story for this](https://www.pivotaltracker.com/story/show/182201617).